### PR TITLE
Add discounts sample app

### PR DESCRIPTION
**Pull request by andrewhassan**
This PR introduces a new Discount Functions Sample App with 3 discount functions (based on the Rust samples). This app is created and the functions are scaffolded with CLI 3.0. The purpose of this app is to serve as an example of how a developer could get up and running with Shopify Functions. This app is meant to be a starting point and does not implement all the functionality a full blown app might have (e.g. Apollo for GraphQL, inline errors on fields, etc.).

**Notes**
1. CLI 3.0 does not have support for pushing functions yet. To work around this, I've added `.shopify-cli.yml` files to each of the functions and pushed them with CLI 2.0. Once 3.0 has support, I will remove these files and update the functions.
2. Right now, shipping discounts don't seem to be working for me. They crash cart/checkout. As such, I wasn't able to 🎩 this, but I can create the discount and see it run successfully in the logs.

**🎩**

1. Ensure you have an app created in Partners.
2. Push the functions to your app. Go into each directory and run `shopify script push` (this should use the old CLI).
3. Ensure that `web/frontend/.env` contains `PRODUCT_DISCOUNT_ID`, `ORDER_DISCOUNT_ID`, and `SHIPPING_DISCOUNT_ID` with their respective function UUIDs.
4. Start the web app with `yarn dev --tunnel`.
5. Go to Discounts in your Admin and try to create/update discounts from this app.
6. Try to checkout as a buyer and see the discounts working.

**Screenshots**

image
image
image

### CODE REVIEW

**Comment by DuncanUszkay1** [comment]
Hm, I think my local environment might be misconfigured for this:
```
Duncans-MacBook-Pro-2:discount-functions-sample-app duncanuszkay$ yarn dev --tunnel
yarn run v1.22.19
$ shopify app dev --tunnel
⭑ You are running a development version of the CLI at:
  /Users/duncanuszkay/src/github.com/Shopify/shopify-cli/bin/shopify

┏━━ Error ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━
┃ The option --tunnel is not supported.
┗━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━
error Command failed with exit code 1.
info Visit https://yarnpkg.com/en/docs/cli/run for documentation about this command.
```
Is `shopify` supposed to point to CLI 3.0 here?

File: sample-apps/discount-functions-sample-app/web/frontend/hooks/useCreateDiscount.js [hunk 1]
```diff
@@ -0,0 +1,45 @@
+import { gql } from 'graphql-request';
+
+import { useShopifyMutation } from './useShopifyMutation';
+
+const CREATE_MUTATION = gql`
+  mutation CreateBundleDiscount($discount: DiscountAutomaticAppInput!) {
```
Comment (dphm) [inline sample-apps/discount-functions-sample-app/web/frontend/hooks/useCreateDiscount.js:6]:
```suggestion
  mutation CreateDiscount($discount: DiscountAutomaticAppInput!) {
```

File: sample-apps/discount-functions-sample-app/web/frontend/hooks/useSavedDiscount.js [hunk 2]
```diff
@@ -0,0 +1,50 @@
+import { useMemo } from 'react';
+import { gql } from 'graphql-request';
+
+import { idToGid } from '../utilities/gid';
+import { useShopifyQuery } from './useShopifyQuery';
+
+const QUERY = gql`
+  query GetBundleDiscount($id: ID!) {
```
Comment (dphm) [inline sample-apps/discount-functions-sample-app/web/frontend/hooks/useSavedDiscount.js:8]:
```suggestion
  query GetDiscount($id: ID!) {
```

File: sample-apps/discount-functions-sample-app/web/frontend/hooks/useSavedDiscount.js [hunk 3]
```diff
@@ -0,0 +1,50 @@
...
+          endsAt
+        }
+      }
+    }
+  }
+`;
+
+export function useSavedDiscount(id) {
+  const { data, isLoading, isError } = useShopifyQuery({
+    key: 'GetBundleDiscount',
```
Comment (dphm) [inline sample-apps/discount-functions-sample-app/web/frontend/hooks/useSavedDiscount.js:24]:
```suggestion
    key: 'GetDiscount',
```

File: sample-apps/discount-functions-sample-app/web/frontend/hooks/useUpdateDiscount.js [hunk 4]
```diff
@@ -0,0 +1,47 @@
+import { gql } from 'graphql-request';
+import { idToGid } from '../utilities/gid';
+
+import { useShopifyMutation } from './useShopifyMutation';
+
+const UPDATE_MUTATION = gql`
+  mutation UpdateBundleDiscount(
```
Comment (dphm) [inline sample-apps/discount-functions-sample-app/web/frontend/hooks/useUpdateDiscount.js:7]:
```suggestion
  mutation UpdateDiscount(
```

File: sample-apps/discount-functions-sample-app/web/frontend/components/function-configuration/ProductDiscount.jsx [hunk 5]
```diff
@@ -0,0 +1,97 @@
...
+import {
+  Stack,
+  TextField,
+  TextContainer,
+  Button,
+  List,
+  TextStyle,
+} from '@shopify/polaris';
+
+import { gidToId } from '../../utilities/gid';
```
Comment (hainsdominic) [inline sample-apps/discount-functions-sample-app/web/frontend/components/function-configuration/ProductDiscount.jsx:12]: Unused import

**Comment by hainsdominic** [comment]
I 🎩ed, feels very nice! had to enable the `extensible_discounts_api_extensions` flag on my app to make it work.

**Pull request merged**